File: ChatServer/hash/hashFunction.js
export function hashFunction(s='') {
    let [a, b, c, d] = [0xa0, 0xb1, 0x11, 0x4d];
    let resultHash= '';

    for (let byte of s.split('')) {
        const tmp = byte.charCodeAt(0);
        a ^= tmp;
        b = b ^ a ^ 0x55;
        c = b ^ 0x94;
        d = c ^ tmp ^ 0x74;
    }

    for (let i of [d, c, a, b]) {
        const tmp = i.toString(16);
        resultHash += tmp.length === 2 ? tmp : `0${tmp}`;
    }

    return resultHash;
}

File: PythonClient/hashes/hash_function.py

def hash_function(s=b''):
    a, b, c, d = 0xa0, 0xb1, 0x11, 0x4d

    for byte in bytearray(s):
        a ^= byte
        b = b ^ a ^ 0x55
        c = b ^ 0x94
        d = c ^ byte ^ 0x74

    return format(d << 24 | c << 16 | a << 8 | b, '08x')
